kyo-prelude/shared/src/main/scala/kyo2/Result.scala

```diff
@@ -1,137 +1,172 @@
                     catch
                         case ex if NonFatal(ex) =>
-                            Failure(ex)
+                            Panic(ex)
 
-        inline def flatten[B](using ev: A <:< Result[B]): Result[B] =
+        def flatten[E2, B](using ev: A <:< Result[E2, B]): Result[E | E2, B] =
             flatMap(ev)
 
-        inline def map[B](inline f: A => B): Result[B] =
+        inline def map[B](inline f: A => B): Result[E, B] =
             flatMap(v => Result.success(f(v)))
 
-        inline def filter(inline p: A => Boolean): Result[A] =
+        inline def withFilter(inline p: A => Boolean): Result[E | NoSuchElementException, A] =
+            filter(p)
+
+        inline def filter(inline p: A => Boolean): Result[E | NoSuchElementException, A] =
             flatMap { v =>
                 if !p(v) then
-                    throw new NoSuchElementException("Predicate does not hold for " + v)
-                v
+                    Error(new NoSuchElementException("Predicate does not hold for " + v))
+                else
+                    v
             }
 
-        inline def recover[B >: A](inline rescueException: PartialFunction[Throwable, B]): Result[B] =
-            (self: @unchecked) match
-                case Failure(exception) if rescueException.isDefinedAt(exception) =>
-                    Result(rescueException(exception))
-                case _ => self.asInstanceOf[Result[B]]
-
-        inline def recoverWith[B >: A](inline rescueException: PartialFunction[Throwable, Result[B]]): Result[B] =
-            (self: @unchecked) match
-                case Failure(exception) if rescueException.isDefinedAt(exception) =>
-                    rescueException(exception)
-                case _ => self.asInstanceOf[Result[B]]
-
-        inline def toEither: Either[Throwable, A] =
-            fold(Left(_))(Right(_))
+        inline def recover[B >: A](inline pf: PartialFunction[Failure[E], B]): Result[E, B] =
+            try
+                (self: @unchecked) match
+                    case self: Failure[E] if pf.isDefinedAt(self) =>
+                        Result.success(pf(self))
+                    case _ => self
+            catch
+                case ex: Throwable if NonFatal(ex) =>
+                    Panic(ex)
+
+        inline def recoverWith[E2, B >: A](inline pf: PartialFunction[Failure[E], Result[E2, B]]): Result[E | E2, B] =
+            try
+                (self: @unchecked) match
+                    case self: Failure[E] if pf.isDefinedAt(self) =>
+                        pf(self)
+                    case _ => self
+            catch
+                case ex: Throwable if NonFatal(ex) =>
+                    Panic(ex)
+
+        def toEither: Either[E | Throwable, A] =
+            fold(Left(_))(Left(_))(Right(_))
+
+        def toTry(using
+            @implicitNotFound("Error type must be a 'Throwable' to invoke 'toTry'. Found: '${E}'")
+            ev: E <:< Throwable
```

Comment:
I assume it's friendlier to consumers to use the implicit vs an extension method limited to `Result[Throwable, A]`?

Reply:
yeah, I don't think it's possible to use `implicitNotFound` if we restrict it via the extension method declaration